From 6579b6c1d014e31ebb139a3ac882bbed6a75923b Mon Sep 17 00:00:00 2001
From: lbarapat <lbarapat@qti.qualcomm.com>
Date: Fri, 14 Aug 2026 16:08:49 +0530
Subject: [PATCH] paho-mqtt-c: add Unix domain socket support to Socket_new

Adds AF_UNIX connection handling to Socket_new so that paho MQTT
clients can connect to unix:// broker URIs. When the stripped server
URI starts with '/' (after the 'unix://' prefix is removed by
MQTTClient.c), Socket_new creates an AF_UNIX SOCK_STREAM socket and
connects to the filesystem path instead of using getaddrinfo.

Required for local GVM services (e.g. c2c_hub pm_a2a_init) that
connect to the internal mosquitto broker via UDS socket at
/data/c2c/conf/a2a_uds.sock.

Signed-off-by: lbarapat <lbarapat@qti.qualcomm.com>
---
 src/Socket.c  | 43 ++++++++++++++++++++++++++++++++++++++++++-
 2 files changed, 43 insertions(+), 1 deletion(-)

diff --git a/src/Socket.c b/src/Socket.c
index bcff03d..db55057 100644
--- a/src/Socket.c
+++ b/src/Socket.c
@@ -43,6 +43,9 @@
 #include <ctype.h>
 
 #include "Heap.h"
+#if !defined(_WIN32) && !defined(_WIN64)
+#include <sys/un.h>
+#endif
 
 #if defined(USE_SELECT)
 int isReady(int socket, fd_set* read_set, fd_set* write_set);
@@ -965,7 +968,7 @@ int Socket_new(const char* addr, size_t addr_len, int port, SOCKET* sock)
 #endif
 {
 	int type = SOCK_STREAM;
-	char *addr_mem;
+	char *addr_mem = NULL;
 	struct sockaddr_in address;
 #if defined(AF_INET6)
 	struct sockaddr_in6 address6;
@@ -981,6 +984,44 @@ int Socket_new(const char* addr, size_t addr_len, int port, SOCKET* sock)
 
 	FUNC_ENTRY;
 	*sock = SOCKET_ERROR;
+
+	/* Unix domain socket: addr starts with '/' after "unix://" prefix stripped */
+#if !defined(_WIN32) && !defined(_WIN64)
+	if (addr[0] == '/')
+	{
+		struct sockaddr_un unix_addr;
+		*sock = socket(AF_UNIX, SOCK_STREAM, 0);
+		if (*sock == SOCKET_ERROR)
+		{
+			rc = SOCKET_ERROR;
+			goto exit;
+		}
+		memset(&unix_addr, 0, sizeof(unix_addr));
+		unix_addr.sun_family = AF_UNIX;
+		strlcpy(unix_addr.sun_path, addr, sizeof(unix_addr.sun_path));
+		if (Socket_addSocket(*sock) == SOCKET_ERROR)
+		{
+			rc = Socket_error("addSocket", *sock);
+			goto exit;
+		}
+		Socket_setnonblocking(*sock);
+		rc = connect(*sock, (struct sockaddr*)&unix_addr, sizeof(unix_addr));
+		if (rc == SOCKET_ERROR)
+		{
+			int err = errno;
+			if (err != EINPROGRESS && err != EWOULDBLOCK)
+			{
+						Socket_close(*sock);
+				*sock = SOCKET_ERROR;
+			}
+			else
+			{
+						rc = 0;
+			}
+		}
+		goto exit;
+	}
+#endif
 	memset(&address6, '\0', sizeof(address6));
 
 	if (addr[0] == '[')
-- 
2.34.1

